doc/sphinx/addendum/type-classes.rst

```diff
@@ -1,578 +1,616 @@
-abbreviate a type, like {\tt relation A := A -> A -> Prop}.
-
-This is equivalent to {\tt Hint Transparent,Opaque} {\ident} {\tt: typeclass\_instances}.
-
-\subsection{\tt Set Typeclasses Axioms Are Instances}
-\optindex{Typeclasses Axioms Are Instances}
-
-This option (off by default since 8.8) automatically declares axioms
-whose type is a typeclass at declaration time as instances of that
-class.
-
-\subsection{\tt Set Typeclasses Dependency Order}
-\optindex{Typeclasses Dependency Order}
-
-This option (on by default since 8.6) respects the dependency order between
-subgoals, meaning that subgoals which are depended on by other subgoals
-come first, while the non-dependent subgoals were put before the
-dependent ones previously (Coq v8.5 and below). This can result in quite
-different performance behaviors of proof search.
-
-\subsection{\tt Set Typeclasses Filtered Unification}
-\optindex{Typeclasses Filtered Unification}
+  (created with Create HintDb for example) with unfoldable variables and
+  constants as the first argument of typeclasses eauto hence makes
+  resolution with the local hypotheses use full conversion during
+  unification.
+
+
+Variants:
+
+#. ``typeclasses eauto [num]``
+
+   *Warning:* The semantics for the limit num
+   is different than for auto. By default, if no limit is given the
+   search is unbounded. Contrary to auto, introduction steps (intro) are
+   counted, which might result in larger limits being necessary when
+   searching with typeclasses eauto than auto.
+
+#. ``typeclasses eauto with {+ @ident}``
+
+   This variant runs resolution with the given hint databases. It treats
+   typeclass subgoals the same as other subgoals (no shelving of
+   non-typeclass goals in particular).
+
+.. _autoapply:
+
+``autoapply term with ident``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The tactic autoapply applies a term using the transparency information
+of the hint database ident, and does *no* typeclass resolution. This can
+be used in ``Hint Extern``’s for typeclass instances (in the hint
+database ``typeclass_instances``) to allow backtracking on the typeclass
+subgoals created by the lemma application, rather than doing type class
+resolution locally at the hint application time.
+
+.. _TypeclassesTransparent:
+
+Typeclasses Transparent, Typclasses Opaque
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. cmd:: Typeclasses { Transparent | Opaque } {+ @ident}
+
+   This commands defines the transparency of the given identifiers
+   during type class resolution. It is useful when some constants
+   prevent some unifications and make resolution fail. It is also useful
+   to declare constants which should never be unfolded during
+   proof-search, like fixpoints or anything which does not look like an
+   abbreviation. This can additionally speed up proof search as the
+   typeclass map can be indexed by such rigid constants (see
+   `8.9.1`_). By default, all constants and local variables are
```

Comment:
TODO ref

Reply:
For this one, if you can't address it now, it is probably a good idea to insert a TODO in the code so that @maximedenes knows it needs to be fixed later.